substrate/frame/staking/src/ledger.rs

```diff
@@ -206,6 +220,19 @@ impl<T: Config> StakingLedger<T> {
 	pub(crate) fn kill(stash: &T::AccountId) -> Result<(), Error<T>> {
 		let controller = <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash)?;
 
+		// call on validator remove or on nominator remove.
+		match crate::Pallet::<T>::status(stash).defensive_unwrap_or(StakerStatus::Idle) {
+			StakerStatus::Validator => <T::EventListeners as OnStakingUpdate<
+				T::AccountId,
+				BalanceOf<T>,
+			>>::on_validator_remove(stash),
+			StakerStatus::Nominator(_) => {
+				let nominations = crate::Pallet::<T>::nominations(stash).unwrap_or_default();
+				<T::EventListeners as OnStakingUpdate<T::AccountId, BalanceOf<T>>>::on_nominator_remove(stash, nominations)
```

Comment:
Should `kill` only kill storage items or do more? It seems `on_nominator_remove` can be called twice when `fn kill_stash` is called, is that expected/ok?

Reply:
The `on_*_remove` won't be called twice, even if the caller has done it already. When the `on_*_remove` are called, the staker status transitions to `Idle`, thus in those cases we skip calling those fn again.

in any case, I have added a defensive message when the staker is not idle by now and more comments.

Reply:
Actually, I changed the logic to return an error + defensive if the stash is not idle when ledger.kill is called. I think this is safer and we enforce the kill to work only if the `Staking::kill_stash` or similar logic has idle'ed the stash before clearing the storage items.